### ISSUE TITLE
@types/parse  Type alias 'Encode' circularly references itself. Type 'Encode' is not generic.

### ISSUE THREAD

**Issue by REPTILEHAUS** [DefinitelyTyped/DefinitelyTyped#42258]
If you know how to fix the issue, make a pull request instead.

- [ X ] I tried using the `@types/parse` package and had problems.
- [ X ] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript

@ullisenmedia

@dpoetzsch 
 @jaeggerr 
@flavionegrao
 @wesleygrimes @owsas @agoldis @AlexandreHetu @dplewis @yomybaby @pocketcolin @rdhelms @jlnquere @yagotome @tybi @RaschidJFR @JeffGuKang @buitanloc @LinusU 

Update parse types just now seems to be issues related to Encode 

```
 node_modules/@types/parse/index.d.ts(501,14): error TS2456: Type alias 'Encode' circularly references itself.
[0] node_modules/@types/parse/index.d.ts(511,41): error TS2315: Type 'Encode' is not generic.
[0] node_modules/@types/parse/index.d.ts(518,29): error TS2315: Type 'Encode' is not generic.
```

**Comment by RaschidJFR** [comment]
There’s been a braking change in #41587 
So you have to update typescript to 3.7.x.

**Comment by REPTILEHAUS** [comment]
As you mentioned here https://github.com/DefinitelyTyped/DefinitelyTyped/pull/41587 this totally breaks my ionic/angular app

**Comment by RaschidJFR** [comment]
You can use [@types/parse@2.10.10](https://www.npmjs.com/package/@types/parse/v/2.10.10), the last version compatible with ts 3.5

**Comment by rdhelms** [comment]
Hi @RaschidJFR and @REPTILEHAUS - thanks for pointing out the issue with Angular's TS support. At the time that we updated to TS 3.7 I wasn't aware of its effect on Angular apps - sorry about that.

How would we like to proceed here? A couple factors to consider:
- Angular 9 is out now with support for TS 3.7, but NOT for TS 3.4 or 3.5: https://github.com/angular/angular/issues/33596
- Older versions of Angular will obviously not support TS 3.7.

At this point I'm inclined to leave it how it is since I'd like to support the most recent version of Angular which at this point is using 3.7...but if you guys think 3.7 is still too recent of a TS version then we could try to look at workarounds.

Off the top of my head I'm not really sure how typing the `toJSON` recursive type would be possible without 3.7's support - but I could take another stab at it.

**Comment by RaschidJFR** [comment]
I haven't taken the time to try to solve it, but in the worst case scenario I think it's better to revert that single feature for now in order to enjoy all other features from the releases after `v2.10.10`.

I'd say even when Angular 9 is out, the ecosystem will take at least 5 months to catch up.

So if you guys are ok with it I propose reverting that change (I can do it if you wish) while figuring out how to make it work in TS 3.5 (or deciding it's time to release it again).

**Comment by rdhelms** [comment]
For clarification, you mentioned "when Angular 9 is out," but it is actually out now, right? So supporting Angular is tricky now because reverting to TS 3.5 will technically be a breaking change for anyone who already started using Angular 9 (since Angular 9 says that they no longer support TS 3.4 or 3.5). But keeping TS 3.7 would continue to break older versions of Angular, which is likely to be the vast majority of Angular users.

I think this may be a case where we should think about implementing separate types files for different versions of TypeScript, as described here: https://github.com/DefinitelyTyped/DefinitelyTyped#i-want-to-use-features-from-typescript-31-or-above

Basically what we're talking about is that TS <3.7 and TS >=3.7 have different type declarations, but we can't just pick one without dropping support for at least one version of Angular. So I think by adding a package.json with `typesVersions` specified, we would be able to support both at the same time.

**Comment by RaschidJFR** [comment]
Wow @rdhelms you've been doing your homework! 😉 That’s a great starting point.
I’m very interested in getting back support for 3.5. I don’t know when I’ll be able to work on this, it looks like a task for just a couple of hours though, so it shouldn’t be that hard. I don’t know about you, but whoever gets a chance just let us know here to see who’s on it.

PS. I meant that Angular 9 is currently in RC14, that’s not the stable 9 version. We can all start playing with it now but it is not officially released. Getting closer everyday.

**Comment by rdhelms** [comment]
I can get started on adding the `typesVersions`, although I will be out of town part of this week. If I don't get it done today then it might not be done until the beginning of next week.

Also, just for my own learnings since I don't live in Angular land as much these days, what's the standard way to know whether Angular 9 is "stable" or not? My impression was that if the docs listed it as "stable" in the corner (which it does now) and if the version surpasses 9.0.0 then that was it - but again I haven't been as close to Angular since v6 or so.
https://github.com/angular/angular/releases

**Comment by RaschidJFR** [comment]
Mmm... so it was actually released 12 days ago. Yes, that's the stable version `9.0.0`. The one before, `9.0.0-rc.14` was the "Release Candidate #14", so it was a prerelease. I was looking to an outdated document which listed RC14 as the latest version, must be updated by now or soon I guess. So you're right, the 9th Coming is here.

**Comment by rdhelms** [comment]
Ok, take a look at https://github.com/DefinitelyTyped/DefinitelyTyped/pull/42459 when you get a chance
